src/charm.py

```diff
@@ -131,20 +134,25 @@ def _reconcile(self) -> None:
         config = CharmConfig.from_charm(self, client)
 
         gateway_resource_definition = GatewayResourceDefinition.from_charm(self)
-        # This line is currently here to validate the existence of the certificates relation.
-        # This charm state component will be used by the upcoming SecretResourceManager.
-        TLSInformation.from_charm(self)
+        tls_information = TLSInformation.from_charm(self)
+        secret_resource_definition = SecretResourceDefinition.from_charm(self)
 
         gateway_resource_manager = GatewayResourceManager(
             labels=self._labels,
             client=client,
         )
+        secret_resource_manager = SecretResourceManager(self._labels, client)
 
         try:
-            gateway = gateway_resource_manager.define_resource(gateway_resource_definition, config)
-        except (CreateGatewayError, InvalidResourceError) as exc:
-            logger.exception("Error creating the gateway resource.")
-            raise RuntimeError("Cannot create gateway.") from exc
+            secret = secret_resource_manager.define_resource(
+                State(secret_resource_definition, config, tls_information)
```

Comment:
Couldn't you create directly a `SecretState` here ? This will avoid using the magic attribute creation in `State`

Reply:
That makes sense, I've also reworked the initialization logic for this class so that it requires exactly those 3 components as parameters.